transaction/core/src/membership_proofs/mod.rs

```diff
@@ -50,6 +50,139 @@ fn hash_nil() -> [u8; 32] {
     hasher.result().try_into().unwrap()
 }
 
+/// Compose two adjacent TxOutMembershipElements into a larger TxOutMembershipElement.
+/// Fails if they are not actually adjacent.
+///
+/// This function can be used to validate a Merkle proof from the bottom up
+/// in a side-channel resistant way. The elements lowest in the tree should be combined,
+/// then the result combined with the next highest element etc. finally getting a range
+/// and hash at the top. This is then checked against the expected hash of the root element.
+///
+/// This function MUST not branch except on erroring code paths
+///
+/// Precondition: a and b are well-formed ranges, a.from <= a.to, b.from <= b.to
+/// It is the callers job to check that.
+///
+/// Arguments:
+/// * a: TxOutMembershipElement to be combined
+/// * b: TxOutMembershipElement to be combined with a, and which should be its left or right sibling
+///
+/// Returns:
+/// * A TxOutMembershipElement for the parent of a and b, with the implied hash value.
+/// * An error if a and b were not adjacent, which implies that the merkle proof was badly structured.
+///   This error can be mapped to e.g. Error::UnexpectedMembershipElement and the caller can provide context.
+pub fn compose_adjacent_membership_elements(
+    a: &TxOutMembershipElement,
+    b: &TxOutMembershipElement,
+) -> Result<TxOutMembershipElement, ()> {
```

Comment:
this is the main ingredient that was missing before to get side-channel resistance